File: backend/src/main/java/com/nusmotion/backend/dto/Building.java
package com.nusmotion.backend.dto;

import com.fasterxml.jackson.annotation.JsonProperty;

/**
 * Maps one building from the NUS Digital Twin API response.
 *
 * LEARNING NOTE:
 * - This record matches the JSON structure returned by the upstream API:
 *   { "elementId": "732229053", "name": "Cendana College",
 *     "address": "28 College Avenue West", "postal": "138533",
 *     "latitude": 1.30814, "longitude": 103.77248 }
 * - The upstream data may have DUPLICATE buildings (same name, slightly different
 *   coordinates).  We handle deduplication in the service layer.
 */
public record Building(
        @JsonProperty("elementId") String elementId,
        @JsonProperty("name") String name,
        @JsonProperty("address") String address,
        @JsonProperty("postal") String postal,
        @JsonProperty("latitude") double latitude,
        @JsonProperty("longitude") double longitude
) {}

File: frontend/lib/data/models/building.dart
class Building {
  final String elementId;
  final String name;
  final String address;
  final String postal;
  final double latitude;
  final double longitude;

  const Building({
    required this.elementId,
    required this.name,
    required this.address,
    required this.postal,
    required this.latitude,
    required this.longitude,
  });

  factory Building.fromJson(Map<String, dynamic> json) => Building(
    elementId: json['elementId'] as String? ?? '',
    name: json['name'] as String? ?? '',
    address: json['address'] as String? ?? '',
    postal: json['postal'] as String? ?? '',
    latitude: (json['latitude'] as num?)?.toDouble() ?? 0,
    longitude: (json['longitude'] as num?)?.toDouble() ?? 0,
  );
}
